src/Aspire.Hosting.AWS.Provisioning/CloudFormation/CloudFormationLifeCycle.cs

```diff
@@ -0,0 +1,237 @@
+                logger.LogInformation("CloudFormation Template for CloudFormation stack {StackName} has not changed", cloudFormationResource.Name);
+            }
+            else
+            {
+                // Update the CloudFormation tag with the latest SHA256.
+                if (shaTag != null)
+                {
+                    shaTag.Value = templateSha256;
+                }
+                else
+                {
+                    tags.Add(new Tag { Key = SHA256_TAG, Value = templateSha256 });
+                }
+
+                var updateStackRequest = new UpdateStackRequest
+                {
+                    StackName = cloudFormationResource.Name,
+                    TemplateBody = templateBody,
+                    Tags = tags
+                };
+
+                logger.LogInformation("Updating CloudFormation stack {StackName}", cloudFormationResource.Name);
+                try
+                {
+                    await cfClient.UpdateStackAsync(updateStackRequest, cancellationToken).ConfigureAwait(false);
+                }
+                catch (AmazonCloudFormationException ex)
+                {
+                    logger.LogError(ex, "Error updating CloudFormation stack {StackName}", cloudFormationResource.Name);
+                    throw new AWSProvisioningException($"Error updating CloudFormation stack {cloudFormationResource.Name}", ex);
+                }
+                
+                stack = await WaitStackToCompleteAsync(cfClient, cloudFormationResource.Name, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        logger.LogDebug("CloudFormation stack has {Count} output parameters", stack.Outputs.Count);
+        if(logger.IsEnabled(LogLevel.Debug))
+        {
+            foreach(var output in stack.Outputs)
+            {
+                logger.LogDebug("Output Name: {Name}, Value {Value}", output.OutputKey, output.OutputValue);
+            }
+        }
+
+        // Capture the CloudFormation stack output parameters on to the Aspire CloudFormation resource. This
+        // allows projects that have a reference to the stack have the output parameters applied to the
+        // projects IConfiguration.
+        cloudFormationResource.Outputs = stack.Outputs;
+    }
+
+    /// <summary>
+    /// Wait for the CloudFormation stack to get to a stable state after creating or updating the stack.
+    /// </summary>
+    /// <param name="cfClient"></param>
+    /// <param name="stackName"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    private async Task<Stack> WaitStackToCompleteAsync(IAmazonCloudFormation cfClient, string stackName, CancellationToken cancellationToken)
+    {
+        const int TIMESTAMP_WIDTH = 20;
+        const int LOGICAL_RESOURCE_WIDTH = 40;
+        const int RESOURCE_STATUS = 40;
+        string mostRecentEventId = string.Empty;
+
+        DateTime minTimeStampForEvents = DateTime.Now;
```

Comment:
`DateTime.Now` or `TimeProvider.GetUtcNow()`

Reply:
I went with `DateTimeOffset.Now` which is what I assume you meant above. There doesn't appear to be a `TimeProvider` injected into the DI by default for an AppHost and I don't want to be the one to introduce it. Admittedly I haven't used `TimeProvider` yet so maybe I misunderstood its usage.

Reply:
I must have had Monday brain as I'd meant `UtcNow` 😅